internal/helm/chart/builder_remote.go

```diff
@@ -17,31 +17,42 @@ limitations under the License.
 package chart
 
 import (
+	"bytes"
 	"context"
 	"fmt"
 	"io"
 	"os"
 	"path/filepath"
 
 	"github.com/Masterminds/semver/v3"
+	"github.com/fluxcd/source-controller/internal/helm/repository"
 	helmchart "helm.sh/helm/v3/pkg/chart"
 	"helm.sh/helm/v3/pkg/chartutil"
+	"helm.sh/helm/v3/pkg/repo"
 	"sigs.k8s.io/yaml"
 
 	"github.com/fluxcd/pkg/runtime/transform"
 
 	"github.com/fluxcd/source-controller/internal/fs"
 	"github.com/fluxcd/source-controller/internal/helm/chart/secureloader"
-	"github.com/fluxcd/source-controller/internal/helm/repository"
 )
 
+// Remote is a repository.ChartRepository or a repository.OCIChartRepository.
+// It is used to download a chart from a remote Helm repository or OCI registry.
+type Remote interface {
+	// GetChart returns a chart.Chart from the remote repository.
+	Get(name, version string) (*repo.ChartVersion, error)
+	// GetChartVersion returns a chart.ChartVersion from the remote repository.
+	DownloadChart(chart *repo.ChartVersion) (*bytes.Buffer, error)
+}
```

Comment:
This interface represents a Repository, maybe it doesn't matter if it's remote or local, download could just be a copy for local repositories. Being in `chart` package, it could be confused with a remote chart due to calling it just Remote.
Based on the implementations of this, it seems to fit better in the `internal/helm/repository` package with a generic name like just `Repository`.
Also, the method descriptions seems to be outdated with some previous names.

We may also need to add an `Unload()` method to this interface. The various implementation can implement it in their own appropriate ways, no-op for some and GC signal for others. I believe this would improve some usage of this interface below for repository download. `downloadFromOCIRepository()` and `downloadFromRepository()` could be unified by moving `StrategicallyLoadIndex()` for non-OCI repository into the implementation of `Get()`. The existing `Get()` can be renamed to something else to not remove the current behavior option if needed. Or the `Get()` interface method could be renamed to `GetChart()` which would be more appropriate as the interface is a repository and GetChart() describes exactly what's being fetched.

Reply:
Remote and local chartBuilder have quite different code base, and a Repository in our code always download from a remote url.

I created the interface here because it is the only place where we need this abstraction for now.

I agree that the name needs to be changed and will propose a new one in the follow-up pr. I will also update the method description, thanks.

I agree to renaming `Get` to `GetChart`, that makes perfect sense. I'm not quite sure about having the `Get` method do more that what it does.

About the Unload() method in the interface, my feeling is that we should stick to having the minimum set of methods in the interface, instead of having no-op implementations.